cpp/include/raft/core/detail/macros.hpp

```diff
@@ -0,0 +1,35 @@
+/*
+ * Copyright (c) 2022, NVIDIA CORPORATION.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+#pragma once
+
+#ifndef _RAFT_HAS_CUDA
+#if defined(__CUDACC__)
+#define _RAFT_HAS_CUDA __CUDACC__
+#endif
+#endif
+
+#ifndef _RAFT_HOST_DEVICE
+#if defined(_RAFT_HAS_CUDA)
+#define _RAFT_HOST_DEVICE __host__ __device__
+#else
+#define _RAFT_HOST_DEVICE
+#endif
+#endif
+
+#ifndef RAFT_INLINE_FUNCTION
+#define RAFT_INLINE_FUNCTION inline _RAFT_HOST_DEVICE
```

Comment:
Was there any particular use-case in mind for this macro where it provides benefit beyond invoking `inline _RAFT_HOST_DEVICE` directly? I could see an argument for introducing a `RAFT_FORCE_INLINE` so that we conditionally make use of the CUDA-specific `__force_inline__`, but it's less obvious to me what the benefit is in tacking on an ordinary `inline` here.

Reply:
This is more for general convenience. It was also being used in our `mdarray` implementation (copied over from `mdspan`, actually) so I figured I'd include it in this refactor (but redefine it as well since we're already redefining the conditional pieces).